test/e2e/init_scale_zero_test.go

```diff
@@ -0,0 +1,313 @@
+	v1a1test "knative.dev/serving/test/v1alpha1"
+)
+
+// TestInitScaleZeroServiceLevel tests setting of annotation checkValidityOnDeploy on
+// the revision level
+func TestInitScaleZeroServiceLevel(t *testing.T) {
+	t.Parallel()
+	cancel := logstream.Start(t)
+	defer cancel()
+
+	clients := Setup(t)
+	names := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "helloworld",
+	}
+
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	t.Log("Creating a new Service with check validity on deploy being false.")
+	objects, _, err := v1a1test.CreateRunLatestServiceReady(t, clients, &names,
+		false, /* https TODO(taragu) turn this on after helloworld test running with https */
+		v1a1testing.WithConfigAnnotations(map[string]string{
+			autoscaling.CheckValidityOnDeployAnnotation: "false",
+		}))
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+
+	// Verify that no pods are created
+	pods := clients.KubeClient.Kube.CoreV1().Pods(test.ServingNamespace)
+	podList, err := pods.List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", serving.RevisionLabelKey, objects.Revision.Name),
+	})
+	if err != nil {
+		t.Fatalf("Failed to list all pods: %v: %v", names.Service, err)
+	}
+	if len(podList.Items) != 0 {
+		t.Fatalf("Did not scale to 0; %d pods are created.", len(podList.Items))
+	}
+
+	t.Log("Creating a new Service with check validity on deploy being true.")
+	names = test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "helloworld",
+	}
+	objects, _, err = v1a1test.CreateRunLatestServiceReady(t, clients, &names,
+		false, /* https TODO(taragu) turn this on after helloworld test running with https */
+		v1a1testing.WithConfigAnnotations(map[string]string{
+			autoscaling.CheckValidityOnDeployAnnotation: "true",
+		}))
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+
+	// Verify that pods are created
+	pods = clients.KubeClient.Kube.CoreV1().Pods(test.ServingNamespace)
+	podList, err = pods.List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", serving.RevisionLabelKey, objects.Revision.Name),
+	})
+	if err != nil {
+		t.Fatalf("Failed to list all pods: %v: %v", names.Service, err)
+	}
+	if len(podList.Items) == 0 {
+		t.Fatal("No pods are created.")
+	}
+}
+
+// TestInitScaleZeroClusterLevel tests setting of annotation checkValidityOnDeploy on
+// the cluster level
+func TestInitScaleZeroClusterLevel(t *testing.T) {
+	cancel := logstream.Start(t)
+	defer cancel()
+
+	clients := Setup(t)
+	namesCheckValidityUnset := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "helloworld",
+	}
+	namesCheckValidityFalse := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "helloworld",
+	}
+	namesCheckValidityTrue := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "helloworld",
+	}
```

Comment:
check my comment below on combining and building up test cases. these code be a part of the buildup too potentially